sw/device/silicon_creator/lib/shutdown.c

```diff
@@ -0,0 +1,277 @@
+      error = e;
+    }
+  }
+#endif
+
+  // For each alert class, configure the various escalation parameters.
+  for (size_t i = 0; i < ALERT_CLASSES; ++i) {
+    config.enabled = enable[i];
+    config.escalation = escalate[i];
+    config.accum_threshold = otp_read32(
+        OTP_CTRL_PARAM_ROM_ALERT_ACCUM_THRESH_OFFSET + i * sizeof(uint32_t));
+    config.timeout_cycles = otp_read32(
+        OTP_CTRL_PARAM_ROM_ALERT_TIMEOUT_CYCLES_OFFSET + i * sizeof(uint32_t));
+    for (size_t phase = 0; phase < ARRAYSIZE(config.phase_cycles); ++phase) {
+      config.phase_cycles[phase] = otp_read32(
+          OTP_CTRL_PARAM_ROM_ALERT_PHASE_CYCLES_OFFSET +
+          (i * ARRAYSIZE(config.phase_cycles) + phase) * sizeof(uint32_t));
+    }
+
+    rom_error_t e = alert_class_configure(kClasses[i], &config);
+    if (e != kErrorOk) {
+      // Keep going if there is an error programming an alert class. We want to
+      // program them all.
+      error = e;
+    }
+  }
+
+  // TODO: initialize interrupt vectors to enter shutdown_finalize.
+  return error;
+}
+
+uint32_t shutdown_redact(rom_error_t reason, shutdown_error_redact_t severity) {
+  uint32_t redacted = (uint32_t)reason;
+  if (reason == kErrorOk) {
+    return 0;
+  }
+  switch (severity) {
+    case kShutdownErrorRedactNone:
+      break;
+    case kShutdownErrorRedactError:
+      redacted &= ~kRomErrorMaskError;
+      break;
+    case kShutdownErrorRedactModule:
+      redacted &= ~(kRomErrorMaskError | kRomErrorMaskModule);
+      break;
+    case kShutdownErrorRedactAll:
+      redacted = kErrorUnknown;
+      break;
+    default:
+      redacted = kErrorUnknown;
+  }
+  return redacted;
+}
+
+SHUTDOWN_FUNC(shutdown_software_escalate(void)) {
+  // FIXME: Use a software alert when available.  For now, signal
+  // a fatal_intg_error from SRAM.
+  enum { kBase = TOP_EARLGREY_SRAM_CTRL_MAIN_BASE_ADDR };
+  abs_mmio_write32(kBase + SRAM_CTRL_ALERT_TEST_REG_OFFSET, 1);
+}
+
+SHUTDOWN_FUNC(shutdown_keymgr_kill(void)) {
+  enum {
+    kBase = TOP_EARLGREY_KEYMGR_BASE_ADDR,
+  };
+  uint32_t reg = bitfield_bit32_write(0, KEYMGR_CONTROL_START_BIT, true);
+  reg = bitfield_field32_write(reg, KEYMGR_CONTROL_DEST_SEL_FIELD,
+                               KEYMGR_CONTROL_DEST_SEL_VALUE_NONE);
+  reg = bitfield_field32_write(reg, KEYMGR_CONTROL_OPERATION_FIELD,
+                               KEYMGR_CONTROL_OPERATION_VALUE_DISABLE);
+  abs_mmio_write32(kBase + KEYMGR_CONTROL_REG_OFFSET, reg);
+  abs_mmio_write32(kBase + KEYMGR_SIDELOAD_CLEAR_REG_OFFSET, 1);
+}
+
+SHUTDOWN_FUNC(shutdown_sram_kill(void)) {
+  enum {
+    kBase = TOP_EARLGREY_SRAM_CTRL_MAIN_BASE_ADDR
+  };
+  // Make sure SRAM exec is disabled and locked.
+  abs_mmio_write32(kBase + SRAM_CTRL_EXEC_REG_OFFSET, 0);
+  abs_mmio_write32(kBase + SRAM_CTRL_EXEC_REGWEN_REG_OFFSET, 0);
+
+  // Request a new scrambling key (making SRAM worthless), then lock.
+  uint32_t reg = bitfield_bit32_write(0, SRAM_CTRL_CTRL_RENEW_SCR_KEY_BIT, true);
```

Comment:
We should probably switch to assembly before doing this. This will e.g. clobber the stack.

I suggest rearranging the calls so that we disable flash before scrambling SRAM and then do the following sequence in a single inline assembly block*:

1. Scramble SRAM
2. Modify ePMP configuration so that only the shutdown function is accessible (could do this before disabling SRAM)
3. Hang

* to avoid the compiler inserting spill handling code etc. between inline assembly blocks

Reply:
Unfortunately Michael is right. In general, compilers assume that you never mess with the stack, so you need to do it in assembly. It's kind of unavoidable.

Reply:
Done...ish.  I've added SRAM kill to the `shutdown_hang` function and left a TODO to add ePMP kill.